rmgpy/solver/base.pyx

```diff
@@ -359,7 +359,39 @@ cdef class ReactionSystem(DASx):
             for l, spec in enumerate(network.source):
                 i = self.get_species_index(spec)
                 self.networkIndices[j,l] = i
-
+                                   
+    @cython.boundscheck(False)                               
+    cpdef maxIndUnderSurfaceLayeringConstraint(self,numpy.ndarray[numpy.float64_t,ndim=1] arr,list surfSpeciesIndices):
+        """
+        determines the index under the surface layering constraint, all of the reactants or all of the products must be in the
+        bulk core (in the core, but not in the surface)
+        """
+        
+        cdef numpy.ndarray[numpy.int_t, ndim=1] sortedIndices
+        cdef int i,j,k,index, numCoreSpecies
+        cdef bool boo
+        
+        numCoreSpecies = self.numCoreSpecies
+        productIndices= self.productIndices
+        reactantIndices = self.reactantIndices
+        
+        sortedIndices = getReverseSortedIndices(arr)
+        
+        for i in xrange(len(sortedIndices)):
+            index = sortedIndices[i]
+            boo = True
+            for j in productIndices[index]:
+                if j in surfSpeciesIndices or j > numCoreSpecies:
+                    boo = False
+            if boo:
+                return index
+            boo = True
+            for j in reactantIndices[index]:
+                if j in surfSpeciesIndices or j > numCoreSpecies:
+                    boo = False
+            if boo:
+                return index
+            
```

Comment:
Is this for loop the same as...?
```
for index in sortedIndices:
    for j in productIndices[index]:
        if j > numCoreSpecies or j in surfSpeciesIndices:
            break
    else: # (didn't break, so all j are in bulk core)
        return True
    for j in reactantIndices[index]:
        if j > numCoreSpecies or j in surfSpeciesIndices:
            break
    else: # (didn't break, so all j are in bulk core)
        return True
```
and is `index` a reaction index but the `j` are species indices?